lib/http_aws_sigv4.c

```diff
@@ -66,6 +70,188 @@ static void sha256_to_hex(char *dst, unsigned char *sha, size_t dst_l)
   }
 }
 
+static char *find_date_hdr(struct Curl_easy *data, const char *sig_hdr)
+{
+  char *tmp = Curl_checkheaders(data, sig_hdr, strlen(sig_hdr));
+
+  if(tmp)
+    return tmp;
+  return Curl_checkheaders(data, STRCONST("Date"));
+}
+
+static CURLcode make_headers(struct Curl_easy *data,
+                             const char *hostname,
+                             char *timestamp,
+                             const char *provider1_low,
+                             const char *provider1_mid,
+                             char **date_header,
+                             struct dynbuf *canonical_headers,
+                             struct dynbuf *signed_headers)
+{
+  char *date_hdr_key =  curl_maprintf("X-%s-Date", provider1_mid);
+  char *date_full_hdr = curl_maprintf("x-%s-date:%s",
+                                      provider1_low, timestamp);
+  struct curl_slist *head = NULL;
+  struct curl_slist *tmp_head = NULL;
+  int ret = CURLE_OUT_OF_MEMORY;
+  struct curl_slist *l;
+  int again = 1;
+  char *full_host = NULL;
+
+  if(!date_hdr_key || !date_full_hdr)
+    goto fail;
+
+  if(Curl_checkheaders(data, STRCONST("Host"))) {
+    head = NULL;
+  }
+  else {
+    if(data->state.aptr.host) {
+      size_t pos;
+
+      full_host = strdup(data->state.aptr.host);
+      if(!full_host)
+        goto fail;
+      /* remove /r/n as the separator for canonical request must be '\n'
+       * and are readd after */
+      pos = strlen(full_host);
+      if(pos > 1 && full_host[pos - 1] == '\n') {
+        full_host[pos - 1] = 0;
+        if(pos > 2 && full_host[pos - 2] == '\r')
+          full_host[pos - 2] = 0;
+      }
+    }
+    else {
+      full_host = curl_maprintf("host:%s", hostname);
```

Comment:
I'm troubled by this code adding its own `Host` header here. It doesn't use nearly the same logic nor does it use the same case or space as the code in http.c. It looks like trouble.

Reply:
It is supposed to be part of so-called 'Canonical Headers'. We need them only to calculate their hash:
https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html
The server will check us by constructing 'Canonical Headers' on its side and calculating the hash. This means that on both sides 'Canonical Headers' must follow the same strict format and order.